File: java/src/main/java/benchmark/Schedule.java
package benchmark;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * B-1 deterministic block_shuffle schedule (must match analysis golden vector).
 *
 * <p>Golden: A,B,C @ seed 42 / message / 1 / abc / bytes / 0 → C,B,A
 * (seed 15992650003647724414).
 */
public final class Schedule {
  private Schedule() {}

  public static String normalizeMode(String mode) {
    String m = mode == null ? "" : mode.trim().toLowerCase(Locale.ROOT);
    if (m.equals("string") || m.equals("buffer")) return "bytes";
    if (m.equals("stream")) return "stream";
    return m;
  }

  /**
   * 64-bit seed for Fisher–Yates for one (cell, mode, rep) group.
   * Returned as a signed long whose bit pattern is the unsigned u64 seed.
   */
  public static long deriveScheduleSeed(
      long baseSeed,
      String typeId,
      int instanceCount,
      String typeConfigHash,
      String mode,
      int rep) {
    String key =
        baseSeed
            + "|"
            + typeId
            + "|"
            + instanceCount
            + "|"
            + (typeConfigHash == null ? "" : typeConfigHash)
            + "|"
            + normalizeMode(mode)
            + "|"
            + rep;
    try {
      MessageDigest md = MessageDigest.getInstance("SHA-256");
      byte[] digest = md.digest(key.getBytes(StandardCharsets.UTF_8));
      // first 8 bytes little-endian
      long u = 0;
      for (int i = 7; i >= 0; i--) {
        u = (u << 8) | (digest[i] & 0xffL);
      }
      return u;
    } catch (NoSuchAlgorithmException e) {
      throw new IllegalStateException("SHA-256 required", e);
    }
  }

  public static <T> List<T> fisherYates(List<T> items, long seed) {
    List<T> arr = new ArrayList<>(items);
    SplitMix64 rng = new SplitMix64(seed);
    for (int i = arr.size() - 1; i > 0; i--) {
      int j = (int) Long.remainderUnsigned(rng.nextU64(), i + 1L);
      T tmp = arr.get(i);
      arr.set(i, arr.get(j));
      arr.set(j, tmp);
    }
    return arr;
  }

  /** Golden vector: A,B,C → C,B,A */
  public static List<String> goldenPermutation() {
    long seed = deriveScheduleSeed(42, "message", 1, "abc", "bytes", 0);
    return fisherYates(List.of("A", "B", "C"), seed);
  }

  public static String resolveStrategy() {
    String env = System.getenv("BENCHMARK_SCHEDULE");
    if (env == null) env = "";
    env = env.trim().toLowerCase(Locale.ROOT);
    if (env.equals("none") || env.equals("block_shuffle")) return env;
    return "block_shuffle";
  }

  public static boolean resolveRecordRunOrder() {
    String env = System.getenv("BENCHMARK_RECORD_RUN_ORDER");
    if (env == null) env = "";
    env = env.trim().toLowerCase(Locale.ROOT);
    if (env.equals("0") || env.equals("false") || env.equals("no")) return false;
    return true;
  }

  /** SplitMix64 PRNG (public-domain algorithm; fixed constants). Unsigned 64-bit via long bits. */
  static final class SplitMix64 {
    private long state;

    SplitMix64(long seed) {
      this.state = seed;
    }

    long nextU64() {
      state += 0x9E3779B97F4A7C15L;
      long z = state;
      z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9L;
      z = (z ^ (z >>> 27)) * 0x94D049BB133111EBL;
      return z ^ (z >>> 31);
    }
  }
}

File: kotlin/src/main/kotlin/benchmark/Schedule.kt
package benchmark

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.Locale

/**
 * B-1 deterministic block_shuffle schedule (must match analysis golden vector).
 *
 * Golden: A,B,C @ seed 42 / message / 1 / abc / bytes / 0 → C,B,A
 */
object Schedule {
    fun normalizeMode(mode: String?): String {
        val m = mode?.trim()?.lowercase(Locale.ROOT) ?: ""
        if (m == "string" || m == "buffer") return "bytes"
        if (m == "stream") return "stream"
        return m
    }

    fun deriveScheduleSeed(
        baseSeed: Long,
        typeId: String,
        instanceCount: Int,
        typeConfigHash: String?,
        mode: String,
        rep: Int,
    ): Long {
        val key =
            "$baseSeed|$typeId|$instanceCount|${typeConfigHash ?: ""}|${normalizeMode(mode)}|$rep"
        val digest = MessageDigest.getInstance("SHA-256").digest(key.toByteArray(StandardCharsets.UTF_8))
        var u = 0L
        for (i in 7 downTo 0) {
            u = (u shl 8) or (digest[i].toLong() and 0xffL)
        }
        return u
    }

    fun <T> fisherYates(items: List<T>, seed: Long): List<T> {
        val arr = items.toMutableList()
        val rng = SplitMix64(seed)
        for (i in arr.lastIndex downTo 1) {
            val j = java.lang.Long.remainderUnsigned(rng.nextU64(), (i + 1).toLong()).toInt()
            val tmp = arr[i]
            arr[i] = arr[j]
            arr[j] = tmp
        }
        return arr
    }

    fun goldenPermutation(): List<String> {
        val seed = deriveScheduleSeed(42, "message", 1, "abc", "bytes", 0)
        return fisherYates(listOf("A", "B", "C"), seed)
    }

    fun resolveStrategy(): String {
        val env = (System.getenv("BENCHMARK_SCHEDULE") ?: "").trim().lowercase(Locale.ROOT)
        if (env == "none" || env == "block_shuffle") return env
        return "block_shuffle"
    }

    fun resolveRecordRunOrder(): Boolean {
        val env = (System.getenv("BENCHMARK_RECORD_RUN_ORDER") ?: "").trim().lowercase(Locale.ROOT)
        return !(env == "0" || env == "false" || env == "no")
    }

    private class SplitMix64(seed: Long) {
        private var state = seed

        fun nextU64(): Long {
            state += C1
            var z = state
            z = (z xor (z ushr 30)) * C2
            z = (z xor (z ushr 27)) * C3
            return z xor (z ushr 31)
        }

        companion object {
            // Same bit patterns as Java's 0x9E3779B97F4A7C15L / 0xBF58476D1CE4E5B9L / 0x94D049BB133111EBL.
            private val C1 = 0x9E3779B97F4A7C15uL.toLong()
            private val C2 = 0xBF58476D1CE4E5B9uL.toLong()
            private val C3 = 0x94D049BB133111EBuL.toLong()
        }
    }
}
